homeassistant/bootstrap.py

```
@@ -890,72 +858,93 @@ async def _async_set_up_integrations(
     watcher = _WatchPendingSetups(hass, _setup_started(hass))
     watcher.async_start()
 
-    domains_to_setup, integration_cache = await _async_resolve_domains_to_setup(
+    integrations, all_integrations = await _async_resolve_domains_and_preload(
         hass, config
     )
-    stage_2_domains = domains_to_setup.copy()
+    all_domains = set(all_integrations)
+    domains = set(integrations)
+
+    _LOGGER.info(
+        "Domains to be set up: %s | %s",
+        domains,
+        all_domains - domains,
+    )
 
     # Initialize recorder
-    if "recorder" in domains_to_setup:
+    if "recorder" in all_domains:
         recorder.async_initialize_recorder(hass)
 
     # Initialize backup
-    if "backup" in domains_to_setup:
+    if "backup" in all_domains:
         backup.async_initialize_backup(hass)
 
-    stage_0_and_1_domains: list[tuple[str, set[str], int | None]] = [
+    stages: list[tuple[str, set[str], int | None]] = [
         *(
-            (name, domain_group & domains_to_setup, timeout)
+            (name, domain_group, timeout)
             for name, domain_group, timeout in STAGE_0_INTEGRATIONS
         ),
-        ("stage 1", STAGE_1_INTEGRATIONS & domains_to_setup, STAGE_1_TIMEOUT),
+        ("1", STAGE_1_INTEGRATIONS, STAGE_1_TIMEOUT),
+        ("2", domains, STAGE_2_TIMEOUT),
     ]
 
-    _LOGGER.info("Setting up stage 0 and 1")
-    for name, domain_group, timeout in stage_0_and_1_domains:
-        if not domain_group:
+    _LOGGER.info("Setting up stage 0")
+    for name, domain_group, timeout in stages:
+        stage_domains_unfiltered = domain_group & all_domains
+        if not stage_domains_unfiltered:
+            _LOGGER.info("Nothing to set up in stage %s: %s", name, domain_group)
+            continue
+
+        stage_domains = stage_domains_unfiltered - hass.config.components
+        if not stage_domains:
+            _LOGGER.info("Already set up stage %s: %s", name, stage_domains_unfiltered)
```

Looks like coverage is missing for this line.  Since its bootstrap, it seems important to cover

---

We should probably never hit this in practice. It's mostly here for the logging, in case we ever do.
Writing a test for it seems quite complex, as we would need to carefully overwrite some STAGE_INTEGRATIONS constants, and doing it just to assert that the log is there, doesn't seem worth it.

---

> We should probably never hit this in practice. It's mostly here for the logging, in case we ever do.
Writing a test for it seems quite complex, as we would need to carefully overwrite some STAGE_INTEGRATIONS constants, and doing it just to assert that the log is there, doesn't seem worth it.

I don't really agree with this sentiment. If we don't think this can ever happen, we should not log at a relatively low level and then silently continue, we should probably raise an error.

What would happen if we dropped the check and tried to set up the domain again?

---

It can happen in theory, so I don't think we should drop the check.

>What would happen if we dropped the check and tried to set up the domain again?

If we dropped the check, then we would just proceed with empty `stage_domains`. I think it would just happen to work, that is it wouldn't do anything, just unnecessary running time. We would get weird looking logs below though.

---

Previously we had these checks in `_async_setup_multi_components` and I don't think they were tested in any way.